src/linalg.jl

```diff
@@ -97,3 +97,194 @@ end
+            ib2 ← B.colptr[j+1]
+            ia ← ia1
+            ib ← ib1
+        end
+        @inbounds for i=1:ia2-ia1+ib2-ib1-1
+            ra ← A.rowval[ia]
+            rb ← B.rowval[ib]
+            if (ra == rb, ~)
+                r += A.nzval[ia]' * B.nzval[ib]
+            end
+            ## b move -> true, a move -> false
+            branch_keeper[i] ⊻= @const ia == ia2-1 || (ib != ib2-1 && ra > rb)
+            ra → A.rowval[ia]
+            rb → B.rowval[ib]
+            if (branch_keeper[i], ~)
+                INC(ib)
+            else
+                INC(ia)
+            end
+        end
+        ~@inbounds for i=1:ia2-ia1+ib2-ib1-1
+            ## b move -> true, a move -> false
+            branch_keeper[i] ⊻= @const ia == ia2-1 || (ib != ib2-1 && A.rowval[ia] > B.rowval[ib])
+            if (branch_keeper[i], ~)
+                INC(ib)
+            else
+                INC(ia)
+            end
+        end
+        ~@routine
+    end
+    ~@routine
+end
+
+# function dot(A::AbstractSparseMatrixCSC{T1,S1},B::AbstractSparseMatrixCSC{T2,S2}) where {T1,T2,S1,S2}
+#     # A and B should be the same size
+#     m, n = size(A)
+#     size(B) == (m,n) || throw(DimensionMismatch("matrices must have the same dimensions"))
+#     r = dot(zero(T1), zero(T2))
+#     # sparse vector dot iterating by columns in A
+#     @inbounds for j = 1:n
+#         # get nonzero element in A[:,j]/B[:,j] with row idx ra/rb 
+#         ia = getcolptr(A)[j]; ia_nxt = getcolptr(A)[j+1]
+#         ib = getcolptr(B)[j]; ib_nxt = getcolptr(B)[j+1]
+#         if ia < ia_nxt && ib < ib_nxt
+#             ra = rowvals(A)[ia]; rb = rowvals(B)[ib]
+#             # iteration in A[:,j]/B[:,j]
+#             while true
+#                 if ra < rb # match false samller ra
+#                     ia += oneunit(S1)
+#                     ia < ia_nxt || break
+#                     ra = rowvals(A)[ia] # update ra bigger
+#                 elseif ra > rb # match false sammler rb
+#                     ib += oneunit(S2)
+#                     ib < ib_nxt || break
+#                     rb = rowvals(B)[ib] # update rb bigger
+#                 else # ra == rb
+#                     r += dot(nonzeros(A)[ia], nonzeros(B)[ib]) # equals to A.nzval[ia]' * B.nzval[ib]
+#                     ia += oneunit(S1); ib += oneunit(S2)
+#                     ia < ia_nxt && ib < ib_nxt || break
+#                     ra = rowvals(A)[ia]; rb = rowvals(B)[ib] # update next ra and rb
+#                 end
+#             end
+#         end
+#     end
+#     return r
+# end
+
+# strange result but I couldn't correct
+@i function idot!(r, x::AbstractVector, A::AbstractSparseMatrix, y::AbstractVector) 
+    @safe length(x) == size(A, 1) || throw(DimensionMismatch())
+    @safe length(y) == size(A, 2) || throw(DimensionMismatch())
+
+    @inbounds for col in size(A, 2)
+        for k in nzrange(A, col)
+            r += (x[A.rowval[k]])' * A.nzval[k] * y[col]
+        end
+    end
+end 
```

Comment:
```suggestion
@i function idot!(r, x::AbstractVector, A::AbstractSparseMatrix{T1}, y::AbstractVector{T2}) where {T1, T2}
    @safe length(x) == size(A, 1) || throw(DimensionMismatch())
    @safe length(y) == size(A, 2) || throw(DimensionMismatch())

    @invcheckoff @inbounds for col in 1:size(A, 2)
        for k in nzrange(A, col)
            @routine begin
                anc ← zero(promote_type(T1,T2))
                anc += A.nzval[k] * y[col]
            end
            r += (x[A.rowval[k]])' * anc
            ~@routine
        end
    end
end 
```

Notes: 
* do not write `y += a * b * c`, currect `PlusEq(*)` only support 3 input arguments, otherwise not differentiable.
* `size(A, 2)` is an integer, it can not be used for iterating.
* this might be not the most performant way of writting it.

Reply:
我去...我调试的时候发现问题在循环...居然没有找出来...我傻了...谢谢学长！